sorted-array based ip_dict

vdimir opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- New IP Dictionary implementation with lower memory consumption, improved performance for some cases, and fixed bugs.

Detailed description / Documentation draft:
- IP Dictionary implementation uses sorted array instead of brtie.
- Fix bugs with wrong IPv4 <-> IPv6 conversions for IP Dictionary. Now it considers IPv6 as IPv4-mapped IPv6-addresses if both types of IPs are presented in dictionary.

vdimir committed:
Use sorted ip array instead of trie in TrieDictionary

vdimir committed:
Add tests/performance/ip_trie.xml [wip]

vdimir committed:
Fix ip subnet comparison

vdimir committed:
Exact ip records comparison for getBlockInputStream

vdimir committed:
[wip] rewrite ip_dict data struct, fix bugs, add tests

vdimir committed:
Fix mappedIPv4ToBinary, add test reference for ip_dict

vdimir commented:
Performance test isn't ready, before it I'd like to implement #16616 to use `SOURCE(CLICKHOUSE(...))`.

vdimir committed:
Minor fixes of ip dict

vdimir committed:
Fix mask comparison for ip dict, add tests

vdimir committed:
SSE version of matchIPv6Subnet for ip_dict

vdimir committed:
Slightly improve ip_dict loading performance, handle v4 to v6 masks in prepossessing, add more tests

vdimir committed:
Add 96 bits to mapped IPv4 in ip_dict loading procedure

vdimir committed:
Remove non-unique entries in ip_dict

vdimir committed:
Minor changes in IP dictionary

KochetovNicolai commented on src/Dictionaries/TrieDictionary.h:
```diff
@@ -150,9 +148,21 @@ class TrieDictionary final : public IDictionaryBase
+    using RowIdxConstIter = ContainerType<size_t>::const_iterator;
+
+    template <typename T> struct IPContainerToValueType {};
+    template <> struct IPContainerToValueType<IPv4Container> { using type = UInt32; };
```
Looks like this code is not valid for gcc: https://stackoverflow.com/a/3052604/6182208
Which causes 
```
../src/Dictionaries/TrieDictionary.h:163:15: error: explicit specialization in non-namespace scope 'class DB::TrieDictionary'
  163 |     template <> struct IPContainerToValueType<IPv4Container> { using type = UInt32; };
```
in ci build

vdimir replied:
I've removed this stuff at all

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+              addr[10] == 0xff && addr[11] == 0xff;
+    if (!success)
+        return 0;
+    return IPv4AsUInt32(&addr[12]);
```
I think strict aliasing rule is violated here, cause we have ptr which points to `uint8_t` and `UInt32` (inside `IPv4AsUInt32`) at the same time.
As I remember, it is ok to use `char` instead of `uint8_t` here.

KochetovNicolai replied:
Also, that's how IPv4ToIPv6 is implemented in another place:
https://github.com/ClickHouse/ClickHouse/blob/a0da588f0608a91240bf1136d955b81d7614fcde/src/Functions/FunctionsCoding.h#L489-L493

I think we just need to use unalignedStore/unalignedLoad functions

KochetovNicolai replied:
Hm, I was wrong about strict aliasing. uint8_t is a typedef for unsigned char, so it is ok. (have mixed it with char8_t)

vdimir replied:
So, can we leave current implementation? This function different from `mapIPv4ToIPv6` (I have `mapIPv4ToIPv6` in this file too, but in other place).

KochetovNicolai replied:
yes

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+                               [&](const auto & a, const auto & b) { return comp(a, b) == 0; });
+    if (new_end != vec.end())
+    {
+        vec.erase(new_end, vec.end());
```
I think we can always call `erase`. Additional check for `new_end == vec.end()` is not needed.

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+    if (new_end != vec.end())
+    {
+        vec.erase(new_end, vec.end());
+        return std::distance(new_end, vec.end());
```
Looks like `new_end` is equals to `vec.end()` after `vec.erase`, so it always return zero

KochetovNicolai replied:
Also dosc say that erase `Invalidates iterators and references at or after the point of the erase`, so `new_end` may be invalid ...

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+}
+
+template <typename T, typename Comp>
+size_t sort_and_unique(std::vector<T> & vec, Comp comp)
```
This name does not fit our style guide

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+    buf[10] = '\xFF';
+    buf[11] = '\xFF';
+    addr = Poco::ByteOrder::toNetwork(addr);
+    memcpy(&buf[12], reinterpret_cast<const uint8_t *>(&addr), 4);
```
I think we don't need reinterpret_cast here. memcpy accept `const void *` anyway

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+
+static bool matchIPv4Subnet(UInt32 target, UInt32 addr, UInt8 prefix)
+{
+    UInt32 mask = (prefix >= 32) ? 0xffffffff : ~(0xffffffff >> prefix);
```
Well, it is almost ok except that `0xffffffff` is an `int` literal, which could be 32 bit for some platforms.
Then,  `0xffffffff` could be signed, and, according to cppreference.com:
```
For signed lhs with nonnegative values, <...> , otherwise the behavior is undefined
```

So, it is better to use `0xffffffffu`

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+#if defined(__SSE2__)
+#include <emmintrin.h>
+
+static bool matchIPv6Subnet(const uint8_t * target, const uint8_t * addr, UInt8 prefix)
```
Maybe it would be easier to store subnet mask as `uint8_t *` (or convert to such representation)
Then we only need to check `(target & mask) == (addr & mask)`

vdimir replied:
Storing only prefix length saves some memory, but it's pretty small, several MB per million of IPs. So It's interesting to compare performance between this approaches in terms of time, I'll try to do it.

KochetovNicolai replied:
Maybe not to store, but convert in place

vdimir replied:
Something like this? It is a bit more effective, but actually this procedure isn't hottest point, so we can choose more readable version.
https://godbolt.org/z/vqrvee

KochetovNicolai replied:
ok

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -331,13 +480,136 @@ void TrieDictionary::loadData()
+                IPAddress addr(addr_str.substr(0, pos));
+                has_ipv6 = has_ipv6 || (addr.family() == Poco::Net::IPAddress::IPv6);
+
+                UInt8 prefix = std::stoi(addr_str.substr(pos + 1), nullptr, 10);
```
Here, I suppose, we should have more user-friendly exception.
stoi may throw `std::invalid_argument`. I suppose, it should be rethrown as `DB::Exception` with key string we tried to parse in message.

Also, string like `127.0.0.1/  24something` will be successfully parsed, which is not ok

KochetovNicolai replied:
The same, probably, should be done if we get `Poco::InvalidAddressException` from `IPAddress` constructor (if we don't have initial key string in message).

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -331,13 +480,136 @@ void TrieDictionary::loadData()
+
+            size_t row_number = ip_records.size();
+
+            std::string addr_str(key_column->getDataAt(row).toString());
```
Extra string copy
Probably, `std::string_view` will work?

vdimir replied:
I cahnge it to `string_view`, but `Poco::Net::IPaddress` accept `const std::string &`, and I convert to string anyway, is it ok?

KochetovNicolai replied:
It is ok. At least it will be copied only for Poco

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -331,13 +480,136 @@ void TrieDictionary::loadData()
+            auto ip_array = IPv6ToBinary(record.addr);
+
+            size_t i = row_idx.size();
+            memcpySmallAllowReadWriteOverflow15(&ipv6_col[i * IPV6_BINARY_LENGTH],
```
I think we can update `IPv6ToBinary` a little bit to write result inplace (to avoid extra copy).
Copy is not expensive here, though  ...

vdimir replied:
I create additional function `IPv6ToRawBinary`

KochetovNicolai commented on src/Dictionaries/TrieDictionary.h:
```diff
@@ -207,32 +217,46 @@ class TrieDictionary final : public IDictionaryBase
 
+    std::variant<IPv4Container, IPv6Container> ip_column;
+    IPMaskContainer mask_column;
+    ContainerType<size_t> parent_subnet;
```
I definitely need comment which explains what is stored here.

Probably, general description about how our keys are stored.

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -331,13 +480,136 @@ void TrieDictionary::loadData()
+    }
+
+    parent_subnet.resize(ip_records.size());
+    std::stack<size_t> subnets_stack;
```
I think we don't even need stack here.
Because it stores values like `i - 1, parent_subnet[i - 1], parent_subnet[parent_subnet[i - 1]], ...` which are accessible all the time.
But this implementation is also ok.

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -485,133 +765,182 @@ TrieDictionary::Attribute TrieDictionary::createAttributeWithType(const Attribut
+
+        const auto & key_mask_column = assert_cast<const ColumnVector<UInt8> &>(*key_columns.back());
+
+        auto comp_v4 = [&](size_t elem, IPv4Subnet target)
```
```suggestion
        auto comp_v4 = [&](size_t elem, const IPv4Subnet & target)
```

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -485,133 +765,182 @@ TrieDictionary::Attribute TrieDictionary::createAttributeWithType(const Attribut
+
+    const auto * key_ip_column_ptr = typeid_cast<const ColumnFixedString *>(&*key_columns.front());
+    if (key_ip_column_ptr == nullptr)
+        throw Exception{"Expected a UInt32 IP column", ErrorCodes::TYPE_MISMATCH};
```
```suggestion
        throw Exception{"Expected a FixedString(16) IP column", ErrorCodes::TYPE_MISMATCH};
```

KochetovNicolai replied:
Also need to check that string size is 16

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -485,133 +765,182 @@ TrieDictionary::Attribute TrieDictionary::createAttributeWithType(const Attribut
+    const auto & key_mask_column = assert_cast<const ColumnVector<UInt8> &>(*key_columns.back());
+
+    const auto * ipv6_col = std::get_if<IPv6Container>(&ip_column);
+    auto comp_v6 = [&](size_t i, IPv6Subnet target)
```
```suggestion
    auto comp_v6 = [&](size_t i, const IPv6Subnet & target)
```

KochetovNicolai commented on src/Dictionaries/TrieDictionary.cpp:
```diff
@@ -485,133 +765,182 @@ TrieDictionary::Attribute TrieDictionary::createAttributeWithType(const Attribut
+            // addrv4 has native endianness
+            auto addrv4 = UInt32(first_column->get64(i));
+            auto found = tryLookupIPv4(addrv4, addrv6_buf);
+            set_value(i, (found != ipNotFound()) ? static_cast<OutputType>(vec[*found]) : get_default(i));
```
It is not very readable. I would prefer `if ... else ...` here

KochetovNicolai reviewed (commented):
Implementation and code is ok in general.
Just some small things need to be fixed.

vdimir committed:
Merge remote-tracking branch 'upstream/master' into ip-dict-no-trie

vdimir commented:
I've tested implementation with RIPE dataset. Data was loaded to ClickHouse table.
Also I generate table with "real" IPs: one random address from each subnet and load into CH too.
Measurements collected with `clickhouse-client -t`. Queries are run 5 times sequentially.

New implementation shows low performance on loading. Also it is worse on querying randomly-generated bits. But it better on "real" addresses. Also estimated memory consumption 40% lower.

`dict_ip_trie` contains ~4M IPv4 subnets
`ip_datasets.ipdata` contains one random ip from each subnet from `dict_ip_trie`

Loading time:
```
trie :  5.709
array: 22.708
```

Query with random data:
```
SELECT COUNT(dictHas('default.dict_ip_trie', tuple(rand32())))
FROM numbers(1000000)
```
```
trie : [0.186, 0.190, 0.212, 0.215, 0.218]
array: [0.379, 0.384, 0.360, 0.348, 0.367]
```

Query with real data from `ipdata`:

```
SELECT COUNT(dictHas('default.dict_ip_trie', tuple(IPv4StringToNum(addr))))
FROM ip_datasets.ipdata
```
```
trie : [1.726, 1.574, 1.815, 1.699, 1.687]
array: [1.126, 1.120, 1.111, 1.064, 1.067]
```

IPv6

`dict_ip6_trie` contains ~1M IPv6 subnets
`ip_datasets.ip6data` contains one random ip from each subnet from `dict_ip6_trie`


Loading time:
```
trie : 2.002
array: 5.785
```

Query with random data:
```
SELECT COUNT(dictHas('default.dict_ip6_trie', tuple(randomFixedString(16))))
FROM numbers(1000000)
```
```
trie : [0.034, 0.045, 0.046, 0.040, 0.037]
array: [0.131, 0.095, 0.097, 0.104, 0.104]
```

Query with real data from `ip6data`:

```
SELECT COUNT(dictHas('default.dict_ip6_trie', tuple(IPv6StringToNum(addr))))
FROM ip_datasets.ip6data
```
```
trie : [0.404, 0.405, 0.369, 0.389, 0.372]
array: [0.282, 0.289, 0.285, 0.281, 0.270]
```


Also I estimate memory usage via `SELECT name, bytes_allocated FROM system.dictionaries`

trie:
````
┌─name──────────┬─bytes_allocated─┐
│ dict_ip_trie  │       570945704 │
│ dict_ip6_trie │       140697768 │
└───────────────┴─────────────────┘
````

array:
```
┌─name──────────┬─bytes_allocated─┐
│ dict_ip_trie  │       356837727 │
│ dict_ip6_trie │        87150902 │
└───────────────┴─────────────────┘
```

vdimir committed:
Fixes for IP dictionary

More meaningful parse errors

Code style fixes, more comments

Fix bytes_allocated calculation

vdimir commented:
@KochetovNicolai Thank you for comments!
I've resolved most of them and leave comments for couple of ones.

vdimir committed:
Remove libbtrie

vdimir committed:
Fix ip dict build

vdimir committed:
Rename TrieDictionary -> IPAddressDictionary

vdimir committed:
Regenerate ya.make, add format null to ip_trie.xml

vdimir commented:
On @alexey-milovidov advice I've tried to use Eytzinger Binary Search https://algorithmica.org/en/eytzinger

It isn't merged to current PR, and I put changes to PR in my fork repository:
https://github.com/vdimir/ClickHouse/pull/1

It performs better on some cases, but case with random generated IPv6 suffers.

Also it requires more time and memory during initialization, arrays constructed and copied several times. Some of this steps could be optimized to eliminate unnecessary copying in loading procedure.

Test is performed in same as above.
```
SELECT 'random IPv4', COUNT(dictHas('default.dict_ip_trie', tuple(rand32()))) FROM numbers(1000000);
SELECT 'real IPv4', COUNT(dictHas('default.dict_ip_trie', tuple(IPv4StringToNum(addr)))) FROM ip_datasets.ipdata;
SELECT 'random IPv6', COUNT(dictHas('default.dict_ip6_trie', tuple(randomFixedString(16)))) FROM numbers(1000000);
SELECT 'real IPv6', COUNT(dictHas('default.dict_ip6_trie', tuple(IPv6StringToNum(addr)))) FROM ip_datasets.ip6data;
```

Current implementation (numbers ~same as in comment above) marked as `current` and Eytzinger layout-based implementation (from separate PR) marked as  `layout`:
```
curent random IPv4 [0.344, 0.311, 0.347, 0.330, 0.298]
layout random IPv4 [0.278, 0.240, 0.212, 0.219, 0.212]

curent real IPv4   [1.126, 1.082, 1.045, 1.106, 1.130]
layout real IPv4   [0.632, 0.539, 0.588, 0.628, 0.566]

curent random IPv6 [0.091, 0.088, 0.089, 0.086, 0.093]
layout random IPv6 [0.293, 0.290, 0.298, 0.297, 0.290]

curent real IPv6   [0.285, 0.269, 0.270, 0.299, 0.267]
layout real IPv6   [0.233, 0.225, 0.254, 0.237, 0.249]
```

KochetovNicolai commented:
Integration test - not connected
Performance - speed up up to 4.8 times:
```
0.566 | 0.116 | -4.858x | -0.795 | 0.786 | ip_trie | 3 | \n SELECT dictGetFloat32(\'default.dict_ip_trie\', \'val\', tuple(IPv6StringToNum(ip)))\n FROM table_ip_from_dict\n WHERE ver == 6\n LIMIT 500000 FORMAT Null\n
-- | -- | -- | -- | -- | -- | -- | --
0.306 | 0.112 | -2.722x | -0.633 | 0.627 | ip_trie | 0 | \n SELECT dictGetFloat32(\'default.dict_ip_trie\', \'val\', tuple(rand32()))\n FROM numbers(500000) FORMAT Null\n
0.317 | 0.175 | -1.809x | -0.448 | 0.437 | ip_trie | 1 | \n SELECT dictGetFloat32(\'default.dict_ip_trie\', \'val\', tuple(randomFixedString(16)))\n FROM numbers(500000) FORMAT Null\n
0.187 | 0.115 | -1.621x | -0.384 | 0.375 | ip_trie | 2 | \n SELECT dictGetFloat32(\'default.dict_ip_trie\', \'val\', tuple(IPv6StringToNum(ip)))\n FROM table_ip_from_dict\n WHERE ver == 4\n LIMIT 500000 FORMAT Null\n
```

Yandex synchronization check - failure in deleted test
```
Exception: 01018_ip_dictionary:                                                    [ FAIL ] - return code 137
E   Received exception from server (version 20.13.1):
E   Code: 137. DB::Exception: Received from localhost:15221. DB::Exception: Unknown dictionary layout type: ip_trie. 
```

Stress test - segfault inside RocsDB
```
2020.11.24 13:12:57.167235 [ 188255 ] {} <Fatal> BaseDaemon: ########################################
2020.11.24 13:12:57.171421 [ 188255 ] {} <Fatal> BaseDaemon: (version 20.13.1.5246, build id: F01B7212634BF42C) (from thread 149516) (query_id: 590d8470-8de9-4043-8d3f-e071c561d877) Received signal Aborted (6)
2020.11.24 13:12:57.171797 [ 188255 ] {} <Fatal> BaseDaemon: 
2020.11.24 13:12:57.171997 [ 188255 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f69e367e3eb 0x7f69e365d899 0x8cc044e 0x3d6afab9 0x3d6af8f3 0x3d5e7231 0x3c39941e 0x3c395e1e 0x3bf26270 0x3bde7161 0x3bddf50a 0x2c094aec 0x2c0947f1 0x2c09adcf 0x2c0b4b43 0x2b82cf1d 0x29c6a731 0x29c6660a 0x29c71198 0x2aee4840 0x2aedf5d5 0x2cc0f2b3 0x2cc3b186 0x37a9640b 0x37a97a2a 0x37f25dd3 0x37f211b9 0x37f1d64c
2020.11.24 13:12:57.175761 [ 188255 ] {} <Fatal> BaseDaemon: 4. raise @ 0x463eb in /usr/lib/x86_64-linux-gnu/libc-2.30.so
2020.11.24 13:12:57.176381 [ 188255 ] {} <Fatal> BaseDaemon: 5. abort @ 0x25899 in /usr/lib/x86_64-linux-gnu/libc-2.30.so
2020.11.24 13:12:57.192369 [ 188255 ] {} <Fatal> BaseDaemon: 6. /build/obj-x86_64-linux-gnu/../base/daemon/BaseDaemon.cpp:384: terminate_handler() @ 0x8cc044e in /usr/bin/clickhouse
2020.11.24 13:12:57.210642 [ 188255 ] {} <Fatal> BaseDaemon: 7. /build/obj-x86_64-linux-gnu/../contrib/libcxxabi/src/cxa_handlers.cpp:61: std::__terminate(void (*)()) @ 0x3d6afab9 in /usr/bin/clickhouse
2020.11.24 13:12:57.220503 [ 188255 ] {} <Fatal> BaseDaemon: 8. /build/obj-x86_64-linux-gnu/../contrib/libcxxabi/src/cxa_handlers.cpp:82: std::terminate() @ 0x3d6af8f3 in /usr/bin/clickhouse
2020.11.24 13:12:57.235931 [ 188255 ] {} <Fatal> BaseDaemon: 9. /build/obj-x86_64-linux-gnu/../contrib/libcxx/src/thread.cpp:47: ? @ 0x3d5e7231 in /usr/bin/clickhouse
2020.11.24 13:12:57.253570 [ 188255 ] {} <Fatal> BaseDaemon: 10. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2363: rocksdb::Timer::Start() @ 0x3c39941e in /usr/bin/clickhouse
2020.11.24 13:12:57.262730 [ 188255 ] {} <Fatal> BaseDaemon: 11. /build/obj-x86_64-linux-gnu/../contrib/rocksdb/db/periodic_work_scheduler.cc:23: rocksdb::PeriodicWorkScheduler::Register(rocksdb::DBImpl*, unsigned int, unsigned int) @ 0x3c395e1e in /usr/bin/clickhouse
2020.11.24 13:12:57.299526 [ 188255 ] {} <Fatal> BaseDaemon: 12. /build/obj-x86_64-linux-gnu/../contrib/rocksdb/db/db_impl/db_impl.cc:702: rocksdb::DBImpl::StartPeriodicWorkScheduler() @ 0x3bf26270 in /usr/bin/clickhouse
2020.11.24 13:12:57.355510 [ 188255 ] {} <Fatal> BaseDaemon: 13. /build/obj-x86_64-linux-gnu/../contrib/rocksdb/db/db_impl/db_impl_open.cc:1811: rocksdb::DBImpl::Open(rocksdb::DBOptions const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<rocksdb::ColumnFamilyDescriptor, std::__1::allocator<rocksdb::ColumnFamilyDescriptor> > const&, std::__1::vector<rocksdb::ColumnFamilyHandle*, std::__1::allocator<rocksdb::ColumnFamilyHandle*> >*, rocksdb::DB**, bool, bool) @ 0x3bde7161 in /usr/bin/clickhouse
2020.11.24 13:12:57.369110 [ 188255 ] {} <Fatal> BaseDaemon: 14. /build/obj-x86_64-linux-gnu/../contrib/rocksdb/db/db_impl/db_impl_open.cc:1465: rocksdb::DB::Open(rocksdb::Options const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, rocksdb::DB**) @ 0x3bddf50a in /usr/bin/clickhouse
2020.11.24 13:12:57.374070 [ 188255 ] {} <Fatal> BaseDaemon: 15. /build/obj-x86_64-linux-gnu/../src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp:0: DB::StorageEmbeddedRocksDB::initDb() @ 0x2c094aec in /usr/bin/clickhouse
2020.11.24 13:12:57.379026 [ 188255 ] {} <Fatal> BaseDaemon: 16. /build/obj-x86_64-linux-gnu/../src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp:256: DB::StorageEmbeddedRocksDB::StorageEmbeddedRocksDB(DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::StorageInMemoryMetadata const&, bool, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x2c0947f1 in /usr/bin/clickhouse
2020.11.24 13:12:57.384659 [ 188255 ] {} <Fatal> BaseDaemon: 17. /build/obj-x86_64-linux-gnu/../base/common/../ext/shared_ptr_helper.h:19: DB::create(DB::StorageFactory::Arguments const&) @ 0x2c09adcf in /usr/bin/clickhouse
2020.11.24 13:12:57.394688 [ 188255 ] {} <Fatal> BaseDaemon: 18. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:1714: std::__1::__function::__func<std::__1::shared_ptr<DB::IStorage> (*)(DB::StorageFactory::Arguments const&), std::__1::allocator<std::__1::shared_ptr<DB::IStorage> (*)(DB::StorageFactory::Arguments const&)>, std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()(DB::StorageFactory::Arguments const&) @ 0x2c0b4b43 in /usr/bin/clickhouse
2020.11.24 13:12:57.398010 [ 188255 ] {} <Fatal> BaseDaemon: 19. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/vector:662: DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, DB::Context&, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const @ 0x2b82cf1d in /usr/bin/clickhouse
2020.11.24 13:12:57.401660 [ 188255 ] {} <Fatal> BaseDaemon: 20. /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterCreateQuery.cpp:786: DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&) @ 0x29c6a731 in /usr/bin/clickhouse
2020.11.24 13:12:57.406171 [ 188255 ] {} <Fatal> BaseDaemon: 21. /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x29c6660a in /usr/bin/clickhouse
2020.11.24 13:12:57.410642 [ 188255 ] {} <Fatal> BaseDaemon: 22. /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::execute() @ 0x29c71198 in /usr/bin/clickhouse
2020.11.24 13:12:57.413806 [ 188255 ] {} <Fatal> BaseDaemon: 23. /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:482: DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*) @ 0x2aee4840 in /usr/bin/clickhouse
2020.11.24 13:12:57.416770 [ 188255 ] {} <Fatal> BaseDaemon: 24. /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:813: DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool) @ 0x2aedf5d5 in /usr/bin/clickhouse
2020.11.24 13:12:57.419821 [ 188255 ] {} <Fatal> BaseDaemon: 25. /build/obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:254: DB::TCPHandler::runImpl() @ 0x2cc0f2b3 in /usr/bin/clickhouse
2020.11.24 13:12:57.425727 [ 188255 ] {} <Fatal> BaseDaemon: 26. /build/obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::run() @ 0x2cc3b186 in /usr/bin/clickhouse
2020.11.24 13:12:57.431435 [ 188255 ] {} <Fatal> BaseDaemon: 27. /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x37a9640b in /usr/bin/clickhouse
2020.11.24 13:12:57.439003 [ 188255 ] {} <Fatal> BaseDaemon: 28. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/atomic:1036: Poco::Net::TCPServerDispatcher::run() @ 0x37a97a2a in /usr/bin/clickhouse
2020.11.24 13:12:57.445596 [ 188255 ] {} <Fatal> BaseDaemon: 29. /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:0: Poco::PooledThread::run() @ 0x37f25dd3 in /usr/bin/clickhouse
2020.11.24 13:12:57.452874 [ 188255 ] {} <Fatal> BaseDaemon: 30. /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:56: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x37f211b9 in /usr/bin/clickhouse
2020.11.24 13:12:57.460258 [ 188255 ] {} <Fatal> BaseDaemon: 31. /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/include/Poco/AutoPtr.h:215: Poco::ThreadImpl::runnableEntry(void*) @ 0x37f1d64c in /usr/bin/clickhouse
``

KochetovNicolai committed:
Update registerDictionaries.cpp

KochetovNicolai committed:
Update arcadia_skip_list.txt

KochetovNicolai committed:
Merge branch 'master' into ip-dict-no-trie

KochetovNicolai commented:
I will merge this pr now, because checks are fine (everything failed is not connected to this pr), and perftests show speedup.
We can add Eytzinger Binary Search in another pr if needed.

alexey-milovidov commented:
@KochetovNicolai @vdimir Let's check what of the issues has been fixed:
#6806
#6804
#1205

vdimir commented:
> @KochetovNicolai @vdimir Let's check what of the issues has been fixed:
I think  #6806 and  #1205 are fixed, but I'm going to check cases from this issues once again and I will add comments to it.
This PR doesn't change #6804, but it seems easy, I'll try to resolve it too.
